oxidation/libparsec_python/src/storage.rs

```diff
@@ -0,0 +1,326 @@
+        };
+        py.allow_threads(|| {
+            self.0
+                .set_manifest(
+                    entry_id.0,
+                    manifest,
+                    cache_only,
+                    removed_ids.map(|x| {
+                        x.into_iter()
+                            .map(|id| libparsec_core_fs::ChunkOrBlockID::ChunkID(id.0))
+                            .collect()
+                    }),
+                )
+                .map_err(|e| PyValueError::new_err(e.to_string()))
+        })
+    }
+
+    fn clear_manifest(&self, py: Python, entry_id: EntryID) -> PyResult<()> {
+        py.allow_threads(|| {
+            self.0
+                .clear_manifest(entry_id.0)
+                .map_err(|_| FSLocalMissError::new_err(entry_id))
+        })
+    }
+
+    fn create_file_descriptor(&self, py: Python, manifest: LocalFileManifest) -> PyResult<u32> {
+        py.allow_threads(|| Ok(self.0.create_file_descriptor(manifest.0).0))
+    }
+
+    fn load_file_descriptor(&self, py: Python, fd: u32) -> PyResult<LocalFileManifest> {
+        py.allow_threads(|| {
+            Ok(LocalFileManifest(
+                self.0
+                    .load_file_descriptor(parsec_api_types::FileDescriptor(fd))
+                    .map_err(|e| FSInvalidFileDescriptor::new_err(e.to_string()))?,
+            ))
+        })
+    }
+
+    fn remove_file_descriptor(&self, py: Python, fd: u32) -> PyResult<()> {
+        py.allow_threads(|| {
+            self.0
+                .remove_file_descriptor(parsec_api_types::FileDescriptor(fd))
+                .map(|_| ())
+                .ok_or_else(|| FSInvalidFileDescriptor::new_err(""))
+        })
+    }
+
+    fn set_clean_block(&self, py: Python, block_id: BlockID, block: &[u8]) -> PyResult<()> {
+        py.allow_threads(|| {
+            self.0
+                .set_clean_block(block_id.0, block)
+                .map_err(|e| PyValueError::new_err(e.to_string()))
+        })
+    }
+
+    fn clear_clean_block(&self, py: Python, block_id: BlockID) -> PyResult<()> {
+        py.allow_threads(|| {
+            self.0.clear_clean_block(block_id.0);
+            Ok(())
+        })
+    }
+
+    fn get_dirty_block<'py>(&self, py: Python<'py>, block_id: BlockID) -> PyResult<&'py PyBytes> {
+        let block = py
+            .allow_threads(|| self.0.get_dirty_block(block_id.0))
+            .map_err(|_| FSLocalMissError::new_err(block_id))?;
+        Ok(PyBytes::new(py, &block))
+    }
+
+    fn get_chunk<'py>(&self, py: Python<'py>, chunk_id: ChunkID) -> PyResult<&'py PyBytes> {
+        let chunk = py
+            .allow_threads(|| self.0.get_chunk(chunk_id.0))
+            .map_err(|_| FSLocalMissError::new_err(chunk_id))?;
+        Ok(PyBytes::new(py, &chunk))
+    }
+
+    // Pyo3 is inefficient with Vec<u8> but set_chunk must handle PyByteArray and PyBytes
+    // but can't safely call unsafe { block.as_bytes() } because it's called in multiple threads
```

Comment:
I think it's more interesting here to keep using the unsafe trick and dropping the use of `py.allow_threads` instead

My guess is `set_chunk()` call should be very fast where the block vector is typically multiple Ko big

Can you run the tests with both approaches to have a rough idea of the perf difference ? (I guess this is going to be unfair for the unsafe approach given the chunks are really small in test, but it may at least show if the copy approach has a very performance drawback)

Reply:
For a block of `4096B`:
```
allow_threads with Vec<u8>: 250µs
unsafe with as_bytes(): 150µs
```
When I move a file in my mountpoint, I see no big difference for a file of 186MB

Reply:
That's intesting ;-)

I guess the copy is okay then !